### PR TITLE
thrift proxy: add support for request mirroring policies

### PR DESCRIPTION
**Pull request by rgs1**
Commit Message: This follows the model used for mirroring HTTP requests.

Additional Description:
This adds supports for shadowing Thrift requests, which is useful for load testing and
testing new services.

Risk Level: Low - new feature.
Testing: Unit tests added.
Docs Changes: Will add.
Release Notes: Added.
Platform Specific Features: NA.

Signed-off-by: Raul Gutierrez Segales <rgs@pinterest.com>

### CODE REVIEW

**Comment by rgs1** [comment]
@zuercher @fishcakez please note: no tests nor docs added yet plus a bunch of things are still missing (stats, more docs, protocol conversion, etc). But it would be nice to get some feedback on the overall approach. Thanks!

**Comment by rgs1** [comment]
/wait 

Feel free to give design feedback, in the meanwhile I'll add stats, tests, docs and handle proto upgrades/conversion.

**Comment by rgs1** [comment]
/wait

Tests coming up in a bit.

**Comment by rgs1** [comment]
/wait

More tests coming up.

**Comment by rgs1** [comment]
/wait

Just fixed a test, didn't push the new ones yet

**Comment by rgs1** [comment]
@zuercher @fishcakez I think this is ready for a first pass – thanks!

File: source/extensions/filters/network/thrift_proxy/router/shadow_writer_impl.cc [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+    first_reply_field_ = false;
+  }
+
+  return FilterStatus::Continue;
+}
+
+FilterStatus NullResponseDecoder::transportEnd() {
+  ASSERT(metadata_ != nullptr);
+  complete_ = true;
+  // TODO: bump stats.
```
Comment (zuercher) [inline source/extensions/filters/network/thrift_proxy/router/shadow_writer_impl.cc:67]: Should this be part of this PR? If not, `// TODO(rgs1): ...`

Comment (rgs1) [reply source/extensions/filters/network/thrift_proxy/router/shadow_writer_impl.cc:67]: That precise comment might be out of date, but there are few other spots. I'll address them, assuming we generally agree with the direction of the PR.

File: source/extensions/filters/network/thrift_proxy/router/shadow_writer_impl.cc [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+    if (!reset_stream_) {
+      request_ready_ = true;
+    }
+  }
+}
+
+// TODO: set response timeout.
+void ShadowRequest::writeRequest() {
+  // TODO(rgs1): is this even needed? The original upstream request does this...
+  metadata_->setProtocol(protocol_->type());
```
Comment (zuercher) [inline source/extensions/filters/network/thrift_proxy/router/shadow_writer_impl.cc:222]: Seems like a race to modify this.

File: source/extensions/filters/network/thrift_proxy/router/shadow_writer_impl.h [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+        upstream_rq_oneway_(stat_name_set_->add("thrift.upstream_rq_oneway")),
+        upstream_rq_invalid_type_(stat_name_set_->add("thrift.upstream_rq_invalid_type")),
+        upstream_resp_reply_(stat_name_set_->add("thrift.upstream_resp_reply")),
+        upstream_resp_reply_success_(stat_name_set_->add("thrift.upstream_resp_success")),
+        upstream_resp_reply_error_(stat_name_set_->add("thrift.upstream_resp_error")),
+        upstream_resp_exception_(stat_name_set_->add("thrift.upstream_resp_exception")),
+        upstream_resp_invalid_type_(stat_name_set_->add("thrift.upstream_resp_invalid_type")),
+        upstream_rq_time_(stat_name_set_->add("thrift.upstream_rq_time")),
+        upstream_rq_size_(stat_name_set_->add("thrift.upstream_rq_size")),
+        upstream_resp_size_(stat_name_set_->add("thrift.upstream_resp_size")) {}
```
Comment (zuercher) [inline source/extensions/filters/network/thrift_proxy/router/shadow_writer_impl.h:142]: Is there a way we can avoid duplicating these here?

**Review by zuercher — COMMENTED** [review]
At a glance, this doesn't look like it supports the shadow cluster using a different protocol from the original cluster. Do I have that right? I expect that's a use case someone will want. Implementing would require separating the the shadow request's transport/protocol/metadata from the original request and I think it probably also means you need to separately negotiate thrift protocol (I assume this version sends the upgrade message if it's send on the main cluster and assumes that it gets the same result as the main cluster).

I think you could probably skip implementing that for now, but in that case we should detect it as a configuration error.

**Comment by zuercher** [comment]
I think perhaps handling conversion and protocol upgrade implies that we should look and see if we can re-use UpstreamRequest for the shadow requests? That way we don't need to duplicate all that for the shadow requests.

**Comment by rgs1** [comment]
> At a glance, this doesn't look like it supports the shadow cluster using a different protocol from the original cluster. Do I have that right? I expect that's a use case someone will want. Implementing would require separating the the shadow request's transport/protocol/metadata from the original request and I think it probably also means you need to separately negotiate thrift protocol (I assume this version sends the upgrade message if it's send on the main cluster and assumes that it gets the same result as the main cluster).
> 
> I think you could probably skip implementing that for now, but in that case we should detect it as a configuration error.

Yup -- I left that aside as I was getting the pieces to fit together. Depending on how much additional work this needs, we might add it now or do it in a 2nd round.

cc: @fishcakez

**Comment by rgs1** [comment]
> I think perhaps handling conversion and protocol upgrade implies that we should look and see if we can re-use UpstreamRequest for the shadow requests? That way we don't need to duplicate all that for the shadow requests.

Maybe, the whole fire and forget part plus the additional changes to decouple from the Router will require an interesting amount of churn... 

That's the reason why I went with making it its own thing. But I do agree, the level of copying isn't optimal thought it's unclear if the right way to address that is by merging this back with UpstreamRequest or just sharing the common code (via a mix-in or common parent).

**Comment by rgs1** [comment]
I am closing this and following up with incremental PRs.

**Pull request closed without merging**